docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,279 @@
+  Admit(a Attributes) (err error)
+  Handles(operation Operation) bool
+}
+```
+
+We'd like to break this into multiple phases, called at the appropriate times
+within the request lifecycle.
+
+```go
+type Interface interface {
+	// ApplyDefaults is executed after the request body is deserialized into an
+	// object for create and update requests. Plugins are free to modify the
+	// object as they see fit. If an error is returned, the request is marked as
+	// failed and no changes are persisted.
+	ApplyDefaults(a Attributes) error
+
+	// Authorize is executed after all plugins' ApplyDefaults methods have
+	// executed. This gives plugins a chance to decide if the request should be
+	// allowed to proceed after the object has had all of its defaults applied. If
+	// an error is returned, the request is marked as failed and no changes are
+	// persisted.
+	Authorize(a Attributes) error
+
+	// Validate is executed just prior to invoking Create, Update, or Delete on
+	// the appropriate `rest.Storage` handler. For Create and Update, Validate is
+	// called after the handler's own `Validate` method is invoked. This gives the
+	// plugin the opportunity to validate the object after the handler has
+	// validated it, but before it is persisted. For Delete, Validate is called
+	// immediately before the handler's `Delete` method is invoked. If an error is
+	// returned, the request is marked as failed and no changes are persisted.
+	Validate(a Attributes) error
+
+  // Handles returns true if this admission controller can handle the given operation
+	// where operation can be one of CREATE, UPDATE, DELETE, or CONNECT
+	Handles(operation Operation) bool
+}
+```
+
+To be able to invoke admission plugin validation at the right time, we'll need
+to modify the various `rest.Storage` interfaces. For example,
+
+```go
+// Creater is an object that can create an instance of a RESTful object.
+type Creater interface {
+  // New returns an empty object that can be used with Create after request data has been put into it.
+  // This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
+  New() runtime.Object
+
+  // Create creates a new version of a resource.
+  Create(ctx api.Context, obj runtime.Object) (runtime.Object, error)
+}
+```
+
+would need to become
+
+```go
+// Creater is an object that can create an instance of a RESTful object.
+type Creater interface {
+  // New returns an empty object that can be used with Create after request data has been put into it.
+  // This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
+  New() runtime.Object
+
+  // BeforeCreate ensures that common operations for all resources are performed
+  // on creation. It is invoked after admission defaulting and authorization. It only returns
+  // errors that can be converted to api.Status. It returns nil if the object should be created.
+  BeforeCreate(ctx api.Context, obj runtime.Object) error
```

Comment:
This probably belongs to the strategy (or should invoke a corresponding method on the strategy).  The strategy is where all business logic on the API objects and their valid transitions lives.  I am coincidentally adding this on the strategy as an optional interface in another change.